### PR TITLE
Switch service controller to shared informers

### PR DESCRIPTION
**Pull request by ncdc**
Originally part of #40097 

cc @deads2k @smarterclayton @gmarek @wojtek-t @timothysc @sttts @liggitt @kubernetes/sig-scalability-pr-reviews

### CODE REVIEW

File: pkg/controller/service/servicecontroller.go [hunk 1]
```diff
@@ -314,6 +319,16 @@ func (s *ServiceController) createLoadBalancerIfNeeded(key string, service *v1.S
 
 func (s *ServiceController) persistUpdate(service *v1.Service) error {
 	var err error
+
```
Comment (ncdc) [inline pkg/controller/service/servicecontroller.go]: Whoops, need to uncomment this - I was testing some stuff

Comment (ncdc) [inline pkg/controller/service/servicecontroller.go]: Fixed

**Comment by ncdc** [comment]
@k8s-bot cvm gke e2e test this

```
W0214 10:02:36.467] ERROR: (gcloud.components.install) The component [alpha] failed to download.
W0214 10:02:36.467] 
W0214 10:02:36.467] ('The read operation timed out',)
```

File: pkg/controller/service/servicecontroller.go [hunk 2]
```diff
@@ -314,6 +319,14 @@ func (s *ServiceController) createLoadBalancerIfNeeded(key string, service *v1.S
...
+	// Make a copy so we don't mutate the shared informer cache
+	copy, err := api.Scheme.DeepCopy(service)
```
Comment (smarterclayton) [inline pkg/controller/service/servicecontroller.go]: Was this actually mutating the cached item?  I expect a TODO for that to not do so.

File: pkg/controller/service/servicecontroller.go [hunk 3]
```diff
@@ -736,30 +749,25 @@ func (s *ServiceController) syncService(key string) error {
...
+	service, err := s.serviceLister.Services(namespace).Get(name)
+	if errors.IsNotFound(err) {
 		// service absence in store means watcher caught the deletion, ensure LB info is cleaned
 		glog.Infof("Service has been deleted %v", key)
 		err, retryDelay = s.processServiceDeletion(key)
+	} else if err != nil {
+		glog.Infof("Unable to retrieve service %v from store: %v", key, err)
+		s.workingQueue.Add(key)
+		return err
 	} else {
```
Comment (smarterclayton) [inline pkg/controller/service/servicecontroller.go]: Don't need an else block here anymore.

Comment (smarterclayton) [inline pkg/controller/service/servicecontroller.go]: Bah, nm.  Ugly construct, use a switch

File: pkg/controller/service/servicecontroller.go (hunk 2 above)
Comment (ncdc) [inline pkg/controller/service/servicecontroller.go]: No, it wasn't. I was working on this kinda late and need to revisit. Will fix.

Comment (ncdc) [inline pkg/controller/service/servicecontroller.go]: But this controller can potentially update `service.Status.LoadBalancer`, so I assume we need a deep copy prior to that happening.

**Comment by ncdc** [comment]
@smarterclayton updated, ptal

File: pkg/controller/service/servicecontroller.go [hunk 4]
```diff
@@ -118,22 +121,12 @@ func New(cloud cloudprovider.Interface, kubeClient clientset.Interface, clusterN
 		cache:            &serviceCache{serviceMap: make(map[string]*cachedService)},
 		eventBroadcaster: broadcaster,
 		eventRecorder:    recorder,
-		nodeLister: listers.StoreToNodeLister{
-			Store: cache.NewStore(cache.MetaNamespaceKeyFunc),
-		},
-		workingQueue: workqueue.NewDelayingQueue(),
+		nodeLister:       nodeInformer.Lister(),
+		nodeListerSynced: nodeInformer.Informer().HasSynced,
+		workingQueue:     workqueue.NewDelayingQueue(),
```
Comment (deads2k) [inline pkg/controller/service/servicecontroller.go]: open an issue to sweep for any unnamed work queue?  Without names we don't get metrics.

Comment (ncdc) [inline pkg/controller/service/servicecontroller.go]: This appears to be the only place that `NewDelayingQueue` or `NewRateLimitingQueue` is used (other than 1 test file).

**Comment by deads2k** [comment]
lgtm, but any bug in any cloudprovider load balancer could bork this, right?

**Comment by ncdc** [comment]
> lgtm, but any bug in any cloudprovider load balancer could bork this, right?

If it modifies the service, you mean?

**Comment by deads2k** [comment]
> If it modifies the service, you mean?

yeah

**Comment by ncdc** [comment]
@deads2k @smarterclayton squashed, updated, and named the queue.

**Comment by ncdc** [comment]
@k8s-bot unit test this

```
I0215 08:47:54.792] test-cmd-util.sh:645: Successful get pod valid-pod {{(index .spec.containers 0).name}}: replaced-k8s-serve-hostname
I0215 08:47:54.879] (BSuccessful
I0215 08:47:54.880] message:error: --grace-period must have --force specified
I0215 08:47:54.880] has:\-\-grace-period must have \-\-force specified
I0215 08:47:54.966] Successful
I0215 08:47:54.966] message:error: --timeout must have --force specified
I0215 08:47:54.966] has:\-\-timeout must have \-\-force specified
I0215 08:47:55.097] node "node-v1-test" created
W0215 08:47:55.198] E0215 08:47:55.097523   18024 actual_state_of_world.go:461] Failed to set statusUpdateNeeded to needed true because nodeName="node-v1-test"  does not exist
W0215 08:47:55.198] E0215 08:47:55.097569   18024 actual_state_of_world.go:475] Failed to update statusUpdateNeeded field in actual state of world: Failed to set statusUpdateNeeded to needed true because nodeName="node-v1-test"  does not exist
W0215 08:47:55.208] E0215 08:47:55.208712   26723 docker_sandbox.go:350] Failed to list checkpoints: failed to list checkpoint store: open /var/lib/dockershim/sandbox: no such file or directory
W0215 08:47:55.432] Error from server (Conflict): error when replacing "STDIN": Operation cannot be fulfilled on nodes "node-v1-test": the object has been modified; please apply your changes to the latest version and try again
```

I believe that points to this line having a problem: https://github.com/kubernetes/kubernetes/blob/beaf5ffacc9b7ede858fc0fcd97c1f0ad27cc58e/hack/make-rules/test-cmd-util.sh#L669

**Comment by ncdc** [comment]
@k8s-bot non-cri e2e test this #37704, #39108

**Pull request merged**